fastestimator/network.py

```diff
@@ -13,51 +13,141 @@
+            else:
+                assert isinstance(op, TensorOp), "unsupported op format, must provide TensorOp in Network"
 
-    def _check_device(self):
-        if self.framework == "pytorch":
-            self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
-            if self.device.type == "cuda":
-                for model in self.model_list:
-                    model.to(self.device)
+    def load_epoch(self, mode, epoch):
+        self.epoch_ops = get_current_ops(self.ops, mode, epoch)
 
-    def prepare(self):
+    def unload_epoch(self):
         pass
 
+    def get_loss_keys(self) -> Set[str]:
+        loss_keys = set()
+        for op in self.ops:
+            if isinstance(op, (RepeatScheduler, EpochScheduler)):
+                for epoch_op in op.get_items():
+                    if isinstance(epoch_op, UpdateOp):
+                        loss_keys.update(to_list(epoch_op.inputs))
+            else:
+                if isinstance(op, UpdateOp):
+                    loss_keys.update(to_list(op.inputs))
+        return loss_keys
+
+    def get_effective_input_keys(self, mode, total_epoches) -> Set[str]:
+        input_keys = set()
+        produced_keys = set()
+        for epoch in self.get_signature_epoches(total_epoches):
+            for op in get_current_ops(self.ops, mode, epoch):
+                input_keys.update(set(key for key in to_list(op.inputs) if key not in produced_keys))
+                produced_keys.update(to_list(op.outputs))
+        input_keys -= {None}
+        return input_keys
+
+    def get_all_output_keys(self, mode, total_epoches) -> Set[str]:
+        output_keys = set()
+        for epoch in self.get_signature_epoches(total_epoches):
+            for op in get_current_ops(self.ops, mode, epoch):
+                output_keys.update(to_list(op.outputs))
+        output_keys -= {None}
+        return output_keys
+
+    def get_signature_epoches(self, total_epoches):
+        signature_epoches = {0}
+        epoch_keys = {0}
+        repeat_cycles = {1}
+        for x in self.ops:
+            if isinstance(x, EpochScheduler):
+                epoch_keys.update(x.epoch_dict.keys())
+            elif isinstance(x, RepeatScheduler):
+                repeat_cycles.add(x.cycle_length)
+        least_common_cycle = lcms(*repeat_cycles)
+        epoch_keys = sorted(epoch_keys)
+        for idx, epoch in enumerate(epoch_keys):
+            if idx + 1 < len(epoch_keys):
+                signature_epoches.update(range(epoch, epoch + min(epoch_keys[idx + 1] - epoch, least_common_cycle)))
+            else:
+                signature_epoches.update(range(epoch, epoch + least_common_cycle))
+        signature_epoches = set(epoch for epoch in signature_epoches if epoch < total_epoches)
+        return signature_epoches
+
+    @staticmethod
+    def _forward_batch(batch: Mapping[str, Any], state: Dict[str, Any], ops: List[TensorOp]):
```

Comment:
any advantage / disadvantage to using System instead of state here?

Reply:
cus tf.function doesn't like object.